keps/sig-api-machinery/2339-storageversion-api-for-ha-api-servers/README.md

```diff
@@ -296,6 +441,204 @@ concern.
+  - kubernetes_healthcheck{name="poststarthook/built-in-resources-storage-version-updater",type="readyz"}
+  - kubernetes_healthcheck{name="poststarthook/built-in-resources-storage-version-updater",type="livez"}
+
+###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?
+
+Not yet, but manual validation should be complete prior to Beta. Steps for manual validation should be reported in this section of the KEP before Beta.
+
+###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?
+
+No
+
+### Monitoring Requirements
+
+###### How can an operator determine if the feature is in use by workloads?
+
+Operators can check the existence of the StorageVersion API, as well as apiserver
+health check metrics to check that the feature is working as expected.
+Operations against the StorageVersion API can be checked via audit logs to see
+if the API is in use by any controllers in the cluster.
+
+###### How can someone using this feature know that it is working for their instance?
+
+- [ ] Events
+  - Event Reason:
+- [X] API .status
+  - Condition name:
+  - Other field: `.status.storageVersions`, `.status.commonEncodingVersion`
+- [X] Other (treat as last resort)
+  - Details: audit logs showing operations against StorageVersion API
+
+###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?
+
+Regarding availability/latency of serving the new API, this should follow the [existing latency SLOs](https://github.com/kubernetes/community/blob/master/sig-scalability/slos/slos.md#steady-state-slisslos)
+for serving mutating or read-only API calls.
+
+Reasonable SLOs specific to the StorageVersion API could be:
+- storageVersion.status.serverStorageVersions is accurately updated for an apiserver within 1 minute of start-up.
+- storageVersion.status.agreedEncodingVersion is accurately updated for the cluster within 1 minute of completing an upgrade.
+
+1 minute seems reasonable since this feature depends on the `APIServerIdentity` feature, that relies on a heart beat from each apiserver
+to create a new Lease object obtaining an ID that will be used in the `storageVersion.status.serverStorageVersions[*].apiServerID` field.
+
+###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?
+
+- [X] Metrics
+  - Metric name: `kubernetes_healthcheck{name="poststarthook/built-in-resources-storage-version-updater"}`
+  - Components exposing the metric: kube-apiserver
+
+###### Are there any missing metrics that would be useful to have to improve observability of this feature?
+
+The following metrics could be useful, but are likely not practical due to cardinality issues or complexity of the implementation:
+- The latency for a single apiserver to update it's encoding version after start-up
```

Comment:
this encoding update is indirectly exposed via `kubernetes_healthcheck{name="poststarthook/built-in-resources-storage-version-updater"}`, right?

Reply:
In some ways yes, although I can imagine scenarios where the time between when the health check passes and when the actual encoding version is updated can be non-negligible.

Reply:
I thought that metric was a simple counter and did not include latency?

Reply:
> I thought that metric was a simple counter and did not include latency?

Given a scrap interval, you can determine how long/often it is false.  The idea for the metrics/sli is an order 5s poll.

Reply:
I still don't think we're missing this, but if you want to leave it in missing, that's ok.